### ISSUE TITLE
"Unreachable definition" warning after infallible expression in `if`-block

### ISSUE THREAD

**Issue by fnetz** [rust-lang/rust#111370]
The following snippet generates an "unreachable definition" warning at `let _x = 1` ([godbolt][1] at 1.69):
```rust
fn infallible() -> std::convert::Infallible {
    loop {}
}

fn main() {
    if false {
        infallible();
    }

    let _x = 1;
}
```

Full output:
```
warning: unreachable definition
  --> <source>:10:9
   |
7  |         infallible();
   |         ------------ any code following this expression is unreachable
...
10 |     let _x = 123;
   |         ^^ unreachable definition
   |
note: this expression has type `Infallible`, which is uninhabited
  --> <source>:7:9
   |
7  |         infallible();
   |         ^^^^^^^^^^^^
   = note: `#[warn(unreachable_code)]` on by default
```

Runtime behaviour is as expected (it exits), so this warning seems wrong.

This occurs in rustc 1.71.0-nightly (c4190f2d3 2023-05-07) all the way back to rustc 1.56.

A similar issue (with an expression instead of a definition) has been brought up in #89779. However, that specific case disappeared in rustc 1.60.

[1]: https://rust.godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:1,endLineNumber:12,positionColumn:1,positionLineNumber:12,selectionStartColumn:1,selectionStartLineNumber:12,startColumn:1,startLineNumber:12),source:'fn+infallible()+-%3E+std::convert::Infallible+%7B%0A++++loop+%7B%7D%0A%7D%0A%0Apub+fn+main()+%7B%0A++++if+false+%7B%0A++++++++infallible()%3B%0A++++%7D%0A%0A++++let+_x+%3D+123%3B%0A%7D%0A'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:output,i:(compilerName:'rustc+1.62.0',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+rustc+1.69.0+(Compiler+%231)',t:'0')),k:50,l:'4',m:57.44920993227991,n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1690,deviceViewOpen:'1',filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'',selection:(endColumn:12,endLineNumber:2,positionColumn:12,positionLineNumber:2,selectionStartColumn:12,selectionStartLineNumber:2,startColumn:12,startLineNumber:2),source:1),l:'5',n:'0',o:'+rustc+1.69.0+(Editor+%231)',t:'0')),header:(),l:'4',m:42.55079006772009,n:'0',o:'',s:0,t:'0')),k:50,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4

**Comment by danjl1100** [comment]
It's interesting that adding an explicit panic satisfies the `unreachable_code` warning:

```patch
     if false {
         infallible();
+        unreachable!();
     }
 
     let _x = 1;
```
 (current stable rustc 1.80.1 on the [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b609e2a01e35f88a1c86c2c38b2e3e72))

**Comment by JonathanBrouwer** [comment]
This is a duplicate of https://github.com/rust-lang/rust/issues/88393

**Comment by JonathanBrouwer** [comment]
@rustbot claim
Together with @bwinterton
